compiler/src/iree/compiler/Dialect/Flow/Transforms/ConvertLinalgMatmulToMmt4D.cpp

```diff
@@ -333,14 +333,16 @@ LinalgMatmulOpToLinalgMmt4DOpPattern::chooseTileParams(Value lhs, Value rhs,
                                   ArrayRef<int> m0k0n0ForMatVec,
                                   std::string comment) {
     assert(m0k0n0ForMatVec[2] == 1 && "not a matrix*vector shape");
-    if (shapeN == 1) {
-      return Mmt4DTileParams(m0k0n0ForMatVec, comment + ", matrix*vector");
-    } else if (shapeM == 1) {
+    if (shapeN != ShapedType::kDynamicSize && shapeN < 4) {
+      int config[3] = {m0k0n0ForMatVec[0], m0k0n0ForMatVec[1],
+                       std::max<int>(shapeN, m0k0n0ForMatVec[2])};
```

Comment:
The implied assumption here is that matrix`*`vector kernel formats can just be widened to other narrow shapes. This may be true in some cases, but won't always be true: sometimes, matrix`*`vector kernels will use very specific shapes that wouldn't work well for any wider right-hand-side.

That is why, just like the existing code allowed to specify entirely independent kernel formats for the general case `m0k0n0` and for the matrix`*`vector case `m0k0n0ForMatVec`, you will need to also allow specifying yet more entirely independent kernel formats for other cases. Naming could become awkward (`m0k0n0ForWhenRhsHas2Columns` ???) so feel free to come up with a larger-scale redesign of this code that makes it better (I don't know right away what such a better design would be but feel free to either come up with your own or schedule a meeting if you would like us to brainstorm together!)

That means that it is important to first know exactly what RHS widths (numbers of columns in the tall, narrow RHS) are important for the benchmarks you're looking at. This PR implemented `N<4`, which includes the rare case `N==3`. Do you care about that? Or do you only care about `N==2`?

The other aspect of this is that codegen needs to generate good code for each of these kernels. At least on ARM, I have found that in many common cases (including the `float32` matrix`*`vector case!) one can't rely on generic `vector.contract` codegen to generate decent code at all, whence the collection of custom inline asm kernels --- which would have to be extended by hand for each new case here, which meshes well with the idea that we have to explicitly list each possible kernel format rather than hope to automatically generalize from the matrix`*`vector case.

Collection of asm kernels:
https://sourcegraph.com/github.com/iree-org/iree@094ec6d183e769893ae6e9ee98e668712a19dd24/-/blob/compiler/src/iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp?L264-684

In particular the comment explaining how generic vector contract codegen botches even the float32 matrix`*`vector case:
https://sourcegraph.com/github.com/iree-org/iree@094ec6d183e769893ae6e9ee98e668712a19dd24/-/blob/compiler/src/iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp?L635-684

Reply:
Thanks for the detailed feedback!

Yes, I only care about N==2 cases, I'll update it to only consider N/M==2 cases. I'll also try the custom kernels, and will let you know if I run into issues! The doc looks really great to me. :D

Reply:
No problem. Don't hesitate if you need help with with ARM assembly or with the particular asm helpers in this file (as you can see, we pre-process the inline asm text so that $(lhs:n) becomes the n-th LHS register, etc. We can go over that in a meeting if that helps.

Reply:
I'm able to generate custom ARM assembly. The below screenshots are derived from Tracy. Would you like to take a quick look if the asm is good? If not, any suggestions here? Thank you!

codegen:

<img width="907" alt="image" src="https://user-images.githubusercontent.com/6949191/187321601-a8c829d9-a13c-444a-9a28-efade5a4fb0f.png">

my custom kernel:

<img width="909" alt="image" src="https://user-images.githubusercontent.com/6949191/187321964-e42cb3fc-c8f4-4d18-b573-83a7ff27c971.png">

Reply:
The codegen's kernel is fine in this case. Let's comment it. Register layout: (v0, v1, v2, v3) are the 8x2 accumulator tile, d4 (first half of v4) is the 2x1 LHS tile, (v6, v7) is the 1x8 RHS tile.

```
add x7, x4, x6, lsl #5 :
    This is doing: x7 = x4 + x6 << 5
    This is computing the pointer to RHS data (from the 512x2048 matrix)
ldr d4, [x10, x6, lsl #3]
    This is loading 2 floats into d4, which is the first 64bit of 128bit register v4, from address x10 + x6 << 3.
    This is loading the LHS data (from the 2x512 matrix)
add x6, x6, #1
    This is doing x6 = x6 + 1, just an increment of the loop counter.
cmp x6, #0x800
    This is comparing x6 to 0x800 = 2048, setting the condition flags for a subsequent conditional branch.
ldp q5, q6, [x7]
    This is loading 256bits of RHS data (from the 512x2048 matrix) from address x7 computed above. The destination registers q5, q6 are the old names for the NEON 128bit registers v5, v6. Some instructions like ldp use the old names of registers.
fmla v1.4s, v5.4s, v4.s[0]
    This is a 4D vector-times-scalar multiply-accumulate, multiplying v5 ( = q5 = RHS data) vector times the first scalar lane of v4 ( = LHS data) accumulating into v1.
...
b.ne .L5
    Conditionally loop, if the above comparison found that x6!=2048
```

So there's not much to say- this is fine.  It's entirely unlike the really bad matrix-vector kernel codegen that I had observed 6 months ago, not sure what makes such a difference, is it the LHS width changing from 1 to 2 or is it the 6 months? Anyway, the codegen being fine, there's no need for a custom asm kernel here (sorry if I caused you to waste time, hope it was still interesting).

Regarding your custom asm kernel, it's mostly equivalent except that

```
mov s5, v4.s[1]
```

is unused --- it is copying `v4.s[1]`, the second lane of v4 (one of the LHS scalars) into `s5`, which is a shorthand name for `v5.s[0]`, the first lane of `v5`, but `v5` is not otherwise used in this kernel. You can drop this from your custom kernel, it shouldn't make a difference, and then it's about exactly similar to the codegen's kernel.

Reply:
Thanks for the deep dive, it's really interesting to me! `mov s5, v4.s[1]` is generated by LLVM backend. I think we don't have much control over it. Anyway, I'll drop the custom asm kernel since it's producing the same code.

I think there are no matvec and vecmat in the model. All the dim sizes are greater than 32. The 20% benchmark regression is a noise in this case.

Reply:
>  Anyway, I'll drop the custom asm kernel since it's producing the same code.

Are we not interested in the i8 case at all , ie. are we not interested in running a quantized flavor of the graph that has the size 2 matmul ?

Reply:
I think eventually we'll be interested in the i8 case. I'm not sure when it will come to our focus. So far, we're more interested in non quantized streaming models. In this case, they are all f32 types.

Reply:
sg. feel free to finalize this PR either way then and i'll approve :-)